src/client/chat.js

```diff
@@ -0,0 +1,287 @@
+        player.lastSignature = currentSignature
+      } else if (player.lastSignature.equals(previousSignature)) {
+        player.lastSignature = currentSignature
+      } else {
+        // Not valid, client can no longer authenticate messages until player quits and reconnects
+        player.hasChainIntegrity = false
+      }
+
+      if (player.hasChainIntegrity) {
+        const verifier = crypto.createVerify('RSA-SHA256')
+        if (previousSignature) verifier.update(previousSignature)
+        verifier.update(concat('UUID', uuid))
+        verifier.update(payload)
+        player.hasChainIntegrity = verifier.verify(player.publicKey, currentSignature)
+      }
+
+      return player.hasChainIntegrity
+    }
+
+    return false
+  }
+
+  client.on('player_info', (packet) => {
+    if (packet.action === 0) { // add player
+      for (const player of packet.data) {
+        if (player.crypto) {
+          client._players[player.UUID] = {
+            publicKey: crypto.createPublicKey({ key: player.crypto.publicKey, format: 'der', type: 'spki' }),
+            publicKeyDER: player.crypto.publicKey,
+            signature: player.crypto.signature,
+            displayName: player.displayName || player.name
+          }
+          client._players[player.UUID].hasChainIntegrity = true
+        }
+      }
+    } else if (packet.action === 4) { // remove player
+      for (const player of packet.data) {
+        delete client._players[player.UUID]
+      }
+    }
+  })
+
+  client.on('message_header', (packet) => {
+    updateAndValidateChat(packet.senderUuid, packet.previousSignature, packet.signature, packet.messageHash)
+
+    client._lastChatHistory.push({
+      previousSignature: packet.previousSignature,
+      signature: packet.signature,
+      messageHash: packet.messageHash
+    })
+  })
+
+  client.on('player_chat', (packet) => {
+    if (!mcData.supportFeature('chainedChatWithHashing')) { // 1.19.0
+      const pubKey = client._players[packet.senderUuid]?.publicKey
+      client.emit('playerChat', {
+        formattedMessage: packet.signedChatContent || packet.unsignedChatContent,
+        type: packet.type,
+        sender: packet.senderUuid,
+        senderName: packet.senderName,
+        senderTeam: packet.senderTeam,
+        verified: pubKey ? client.verifyMessage(pubKey, packet) : false
+      })
+      return
+    }
+
+    const hash = crypto.createHash('sha256')
+    hash.update(concat('i64', packet.salt, 'i64', packet.timestamp / 1000n, 'pstring', packet.plainMessage, 'i8', 70))
+    if (packet.formattedMessage) hash.update(packet.formattedMessage)
+    for (const previousMessage of packet.previousMessages) {
+      hash.update(concat('i8', 70, 'UUID', previousMessage.messageSender))
+      hash.update(previousMessage.messageSignature)
+    }
+
+    // Chain integrity remains even if message is considered unverified due to expiry
+    const tsDelta = Date.now() - packet.timestamp
```

Comment:
packet.timestamp is a bigint so Date.now must be wrapped in the BigInt constructor